server/internal/server.go

```diff
@@ -105,3 +98,27 @@ func (s *Server[Config]) validateAuthenticationToken(ctx context.Context) error
 
 	return status.Error(codes.Unauthenticated, "invalid or missing token")
 }
+
+func RegisterAdapter[Config any](s *Server, datasourceType string, adapter framework.Adapter[Config]) error {
+	// Check for duplicate datasource types
+	if _, ok := s.AdapterGetPageFuncs[datasourceType]; ok {
+		return fmt.Errorf("duplicate datasource type provided: %s", datasourceType)
+	}
+
+	s.AdapterGetPageFuncs[datasourceType] = func(ctx context.Context, req *api_adapter_v1.GetPageRequest) (framework.Response, *entityReverseIdMapping) {
+		adapterRequest, reverseMapping, adapterErr := getAdapterRequest[Config](req)
+		if adapterErr != nil {
+			adapterErrRetryAfter := adapterErr.RetryAfter.AsDuration()
+
+			return framework.NewGetPageResponseError(&framework.Error{
+				Message:    adapterErr.Message,
+				Code:       adapterErr.Code,
+				RetryAfter: &adapterErrRetryAfter,
+			}), nil
+		}
+
+		return adapter.GetPage(ctx, adapterRequest), reverseMapping
+	}
+
+	return nil
+}
```

Comment:
I can't seem to find a clean way to expose this function from the `server` pkg without also having to expose and/or move `entityReverseIdMapping` and `getAdapterRequest` due to import cycles.

The `server` pkg needs to import `internal` to access the `Server` struct defined in this file. Moving the `Server` struct to the `server` pkg has the same issue, as we'd need to also move `AdapterGetPageFunc` which in turn relies on `entityReverseIdMapping`.

The `internal` pkg needs to use `RegisterAdapter` for `TestServer_GetPage`. I could move `RegisterAdapter` and `TestServer_GetPage` to the `server` pkg, but this would still require converting `entityReverseIdMapping` and `getAdapterRequest` to public types/functions.

@rlenglet In your opinion, would you rather require callers of this function to have to specify `internal.RegisterAdapter(...)` or expose `entityReverseIdMapping` and `getAdapterRequest` from the internal package?

Reply:
Nothing from an `internal` package can be exposed.
But anything in the parent package can access exported symbols from an `internal` sub-package.
The only way to solve that is to implement a wrapper `RegisterAdapter` function in the parent `server` package that calls the internal one. E.g. in `server/server.go`, add:
```
func RegisterAdapter[Config any](s *api_adapter_v1.AdapterServer, datasourceType string, adapter Adapter[Config]) error {
	internalServer, ok := s.(*internal.Server)
	if !ok {
		return errors.New("...")
	}
	return internal.RegisterAdapter(internalServer, datasourceType, adapter)
}
```

Reply:
Ahh, I didn't realize go has special treatment for `internal` packages. Added the following above